meadow-client/src/Halogen/Blockly.purs

```diff
@@ -0,0 +1,130 @@
+import Data.Symbol (SProxy(..))
+import Effect.Class (class MonadEffect)
+import Effect.Class.Console (log)
+import Halogen (ClassName(..), Component, ComponentDSL, ComponentHTML, RefLabel(..), action, get, lifecycleComponent, liftEffect, modify, raise)
+import Halogen.HTML (button, div, text)
+import Halogen.HTML as HH
+import Halogen.HTML.Events (input_, onClick)
+import Halogen.HTML.Events as HE
+import Halogen.HTML.Properties (classes, id_, ref)
+import Marlowe.Blockly (buildBlocks, buildGenerator)
+import Marlowe.Parser as Parser
+import Marlowe.Pretty (pretty)
+import Marlowe.Types (Contract(..))
+import Prelude (Unit, bind, discard, pure, show, unit, ($))
+import Text.Parsing.Parser (runParser)
+import Text.Parsing.Parser.Basic (parens)
+
+type BlocklyState
+  = {blocklyState :: Maybe BT.BlocklyState, generator :: Maybe Generator}
+
+_blocklyState :: Lens' BlocklyState (Maybe BT.BlocklyState)
+_blocklyState = prop (SProxy :: SProxy "blocklyState")
+
+_generator :: Lens' BlocklyState (Maybe Generator)
+_generator = prop (SProxy :: SProxy "generator")
+
+data BlocklyQuery a
+  = Inject (Array BlockDefinition) a
+  | Resize a
+  | SetData Unit a
+  | GetCode a
+  | SetCode String a
+
+data BlocklyMessage
+  = Initialized
+  | CurrentCode String
+
+type HTML
+  = ComponentHTML BlocklyQuery
+
+type DSL m
+  = ComponentDSL BlocklyState BlocklyQuery BlocklyMessage m
+
+blockly :: forall m. MonadEffect m => Array BlockDefinition -> Component HH.HTML BlocklyQuery Unit BlocklyMessage m
+blockly blockDefinitions =
+  lifecycleComponent
+    { initialState: \_ -> {blocklyState: Nothing, generator: Nothing}
+    , render
+    , eval
+    , initializer: Just $ action (Inject blockDefinitions)
+    , finalizer: Nothing
+    , receiver: HE.input SetData
+    }
+
+eval :: forall m. MonadEffect m => BlocklyQuery ~> DSL m
+eval (Inject blockDefinitions next) = do
+  blocklyState <- liftEffect $ Blockly.createBlocklyInstance (ElementId "blocklyWorkspace") (ElementId "blocklyToolbox")
+  liftEffect $ Blockly.addBlockTypes blocklyState blockDefinitions
+  let _ = ST.run (do
+                    workspaceRef <- STRef.new blocklyState.workspace
+                    Blockly.initializeWorkspace blocklyState.blockly workspaceRef)
+  let generator = buildGenerator blocklyState
+  _ <- modify _ {blocklyState = Just blocklyState, generator = Just generator}
+  pure next
+
+eval (Resize next) = do
+  state <- get
+  _ <- liftEffect $ Blockly.resize state.blocklyState
+  pure next
+
+eval (SetData _ next) = pure next
+
+eval (GetCode next) = runMaybeT f *> pure next
+  where
+  f = do
+    blocklyState <- MaybeT $ use _blocklyState
+    generator <- MaybeT $ use _generator
+    let code = workspaceToCode blocklyState generator
+        result = runParser code (parens Parser.contract <|> Parser.contract)
+    lift $ case result of
+        Left e -> log (show e)
+        Right contract -> raise $ CurrentCode (show (pretty contract))
+
```

Comment:
Another problem (which was also in the original Meadow): If there are loose contract pieces they probably should be ignored.
Maybe these two reviews should be tickets instead.

Reply:
I'm not sure if it's possible, `workspaceToCode` fails

Reply:
ok so now if there are loose pieces I report an error to the user "This workspace cannot be converted to code"

Reply:
In the deployed version, if I convert Escrow to blockly, then add a loose Null on top and then convert back, I get Null as the whole contract.

Reply:
weird, it seems that it makes a difference _where_ you place that null. The thing is I don't think there's anything we can do about this since it all works in blockly. Outside of blockly we have no idea if there are loose blocks so we cannot act on them.